src/omv/imlib/imlib.h

```diff
@@ -336,75 +336,134 @@ extern const uint16_t ironbow_table[256];
+    PIXFORMAT_BAYER_GRBG = (PIXFORMAT_FLAGS_CR  | (PIXFORMAT_ID_BAYER_GRBG << PIXFORMAT_BPP_BITS) | PIXFORMAT_BPP_BAYER    ),
+    PIXFORMAT_BAYER_RGGB = (PIXFORMAT_FLAGS_CR  | (PIXFORMAT_ID_BAYER_RGGB << PIXFORMAT_BPP_BITS) | PIXFORMAT_BPP_BAYER    ),
+    PIXFORMAT_JPEG       = (PIXFORMAT_FLAGS_J   | (PIXFORMAT_ID_JPEG       << PIXFORMAT_BPP_BITS) | 0                      ),
+    PIXFORMAT_LAST       = 0xFFFFFFFFU,
 } pixformat_t;
 
-typedef enum image_bpp
-{
-    IMAGE_BPP_BINARY,       // BPP = 0
-    IMAGE_BPP_GRAYSCALE,    // BPP = 1
-    IMAGE_BPP_RGB565,       // BPP = 2
-    IMAGE_BPP_BAYER,        // BPP = 3
-    IMAGE_BPP_JPEG          // BPP > 3
+#define PIXFORMAT_BAYER_ANY             \
+        PIXFORMAT_BAYER_BGGR:           \
+        case PIXFORMAT_BAYER_GBRG:      \
+        case PIXFORMAT_BAYER_GRBG:      \
+        case PIXFORMAT_BAYER_RGGB       \
+
+#define PIXFORMAT_IS_VALID(x)           \
+    ((x == PIXFORMAT_BINARY)            \
+     || (x == PIXFORMAT_GRAYSCALE)      \
+     || (x == PIXFORMAT_RGB565)         \
+     || (x == PIXFORMAT_BAYER)          \
+     || (x == PIXFORMAT_BAYER_BGGR)     \
+     || (x == PIXFORMAT_BAYER_GBRG)     \
+     || (x == PIXFORMAT_BAYER_GRBG)     \
+     || (x == PIXFORMAT_BAYER_RGGB)     \
+     || (x == PIXFORMAT_JPEG))          \
+
+#if defined(__BYTE_ORDER__) && (__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__)
+#define PIXFORMAT_STRUCT            \
+struct {                            \
+  union {                           \
+    struct {                        \
+        uint32_t bpp            :8; \
+        uint32_t pixfmt_id      :8; \
+        uint32_t /*reserved*/   :8; \
+        uint32_t is_bayer       :1; \
+        uint32_t is_compressed  :1; \
+        uint32_t is_color       :1; \
+        uint32_t is_mutable     :1; \
+        uint32_t /*reserved*/   :4; \
+    };                              \
+    uint32_t pixfmt;                \
+  };                                \
+  uint32_t size; /* for compressed images */ \
+}
+#elif defined(__BYTE_ORDER__) && (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__)
+#define PIXFORMAT_STRUCT            \
+struct {                            \
+  union {                           \
+    struct {                        \
+        uint32_t /*reserved*/   :4; \
+        uint32_t is_mutable     :1; \
+        uint32_t is_color       :1; \
+        uint32_t is_compressed  :1; \
+        uint32_t is_bayer       :1; \
+        uint32_t /*reserved*/   :8; \
+        uint32_t pixfmt_id      :8; \
+        uint32_t bpp            :8; \
+    };                              \
+    uint32_t pixfmt;                \
+  };                                \
+  uint32_t size; /* for compressed images */ \
 }
-image_bpp_t;
+#else
+#error "Byte order is not defined."
+#endif
 
 typedef struct image {
-    int w;
-    int h;
-    int bpp;
+    uint32_t w;
+    uint32_t h;
+    PIXFORMAT_STRUCT;
     union {
         uint8_t *pixels;
```

Comment:
Can you do another PR to just pick pixels/data as the pointer versus having both?

Reply:
Yes, open an issue.